File: Question 0022 - Generate Parentheses/src/Solution.java
import java.util.ArrayList;
import java.util.List;

class Solution
{
    private void solve(List<String> ans, StringBuilder parenthesis, int opening, int closing)
    {
        //base case
        if(opening == 0 && closing == 0)
        {
            ans.add(parenthesis.toString());
            return;
        }

        if(opening > 0)
        {
            parenthesis.append("(");
            solve(ans,parenthesis,opening-1,closing);
            parenthesis.deleteCharAt(parenthesis.length()-1);
        }
        if(closing > opening)
        {
            parenthesis.append(")");
            solve(ans,parenthesis,opening,closing-1);
            parenthesis.deleteCharAt(parenthesis.length()-1);
        }


    }
    public List<String> generateParenthesis(int n)
    {
        List<String> ans = new ArrayList<>();
        StringBuilder parenthesis = new StringBuilder();
        solve(ans,parenthesis,n,n);
        return ans;
    }
}

File: Question 0022 - Generate Parentheses/src/Solution.cpp
#include<bits/stdc++.h>
using namespace std;

void solve(vector<string> &ans, string &curr, int start, int close)
{
    if(start == 0 && close == 0)
    {
        ans.push_back(curr);
        return;
    }

    if(start > 0)
    {
        curr += "(";
        solve(ans, curr, start-1, close);
        curr.pop_back();
    }
    if(close > start)
    {
        curr += ")";
        solve(ans, curr, start, close-1);
        curr.pop_back();
    }
}

class Solution 
{
public:
    vector<string> generateParenthesis(int n) 
    {
        vector<string> ans;
        string curr = "";
        solve(ans, curr, n, n);
        return ans;
    }
};